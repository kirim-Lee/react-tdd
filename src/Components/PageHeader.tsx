import { useMemo } from 'react';
import { Link, useLocation } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #1e40ff;
`;

const Title = styled.div`
  padding: 20px;
  color: #fff;
  font-size: 20px;
  font-weight: 600;
`;

const GoBack = styled(Link)`
  padding: 20px;
  color: #fff;
  font-size: 16px;
  font-weight: 600;
  text-decoration: none;
  position: absolute;
  left: 20px;
`;

const PageHeader = () => {
  const { pathname } = useLocation();

  const title = useMemo(() => {
    if (pathname === '/') return '할 일 목록';
    if (pathname === '/add') return '할 일 추가';
    if (pathname.startsWith('/detail')) return '할 일 상세';
    return '에러';
  }, [pathname]);

  return (
    <Container>
      <Title>{title}</Title>
      {pathname !== '/' && <GoBack to="/">돌아가기</GoBack>}
    </Container>
  );
};
export default PageHeader;
