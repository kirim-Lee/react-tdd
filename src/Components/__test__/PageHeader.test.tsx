import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import PageHeader from '../PageHeader';

describe('<PageHeader />', () => {
  it('렌더 테스트', () => {
    const router = createMemoryRouter(
      [{ path: '*', element: <PageHeader /> }],
      { initialEntries: ['/'], initialIndex: 0 }
    );

    const { container } = render(<RouterProvider router={router} />);

    const label = screen.getByText('할 일 목록');
    expect(label).toBeInTheDocument();

    const goBack = screen.queryByText('돌아가기');
    expect(goBack).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('/add 로 접속했을때 돌아가기 링크표기와 올바른 타이틀 표기 확인', () => {
    const router = createMemoryRouter(
      [{ path: '*', element: <PageHeader /> }],
      { initialEntries: ['/'], initialIndex: 0 }
    );

    router.navigate('/add');

    render(<RouterProvider router={router} />);

    const label = screen.getByText('할 일 추가');
    expect(label).toBeInTheDocument();

    const goBack = screen.getByText('돌아가기');
    expect(goBack).toBeInTheDocument();
  });

  it('/detail/:id 로 접속했을 때 렌더링 잘 되는지 확인', () => {
    const router = createMemoryRouter(
      [{ path: '*', element: <PageHeader /> }],
      { initialEntries: ['/detail/1'], initialIndex: 0 }
    );

    render(<RouterProvider router={router} />);

    const label = screen.getByText('할 일 상세');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('돌아가기');
    expect(goBack).toBeInTheDocument();
    expect(goBack.getAttribute('href')).toBe('/');
  });

  it('정의되지 않은 경로는 에러로 타이틀 노출되는지 확인', () => {
    const router = createMemoryRouter(
      [{ path: '*', element: <PageHeader /> }],
      { initialEntries: ['/badUrl'], initialIndex: 0 }
    );

    render(<RouterProvider router={router} />);

    const label = screen.getByText('에러');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('돌아가기');
    expect(goBack).toBeInTheDocument();
    expect(goBack.getAttribute('href')).toBe('/');
  });

  it('돌아가기 버튼 눌렀을 때 홈으로 잘 돌아가는지 확인', async () => {
    const router = createMemoryRouter(
      [{ path: '*', element: <PageHeader /> }],
      { initialEntries: ['/badUrl'], initialIndex: 0 }
    );

    render(<RouterProvider router={router} />);

    const goBack = screen.getByText('돌아가기');

    fireEvent.click(goBack);

    const label = screen.getByText('할 일 목록');
    expect(label).toBeInTheDocument();
    expect(goBack).not.toBeInTheDocument();
  });
});
