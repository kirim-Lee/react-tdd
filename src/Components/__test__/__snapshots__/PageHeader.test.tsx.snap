// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader /> 렌더 테스트 1`] = `
<div>
  <div
    class="sc-bcXHqe kNccn"
  >
    <div
      class="sc-gswNZR gUIhpX"
    >
      할 일 목록
    </div>
  </div>
</div>
`;
